In [1]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
from sqlalchemy import create_engine

In [2]:
# load data as dataframe from sql
engine = create_engine('postgresql://postgres:postgres@localhost:5432/Nashville')
violation_types_df = pd.read_sql_query("SELECT * FROM violation_codes", engine)
# visually inspect
violation_types_df.head(5)

violation_id org_id violation_type                  violation_desc  \
0            1      1        BANNERS                         Banners   
1            2      1     BGMATOWNER  Proper Maintenance Req - Owner   
2            4      1      BLDGMAINT          Proper Maintenance Req   
3            5      1     BLDGPERMIT        Building Permit Required   
4            6      1      BLDGSCOPE          Scope of Building Code   

                                      violation_text     remedial_text  \
0  Section 17.32.060 & 17.32.070 -  Banners: Bann...      CAAZ_BANNERS   
1  Section 16.16.030 (B) - Proper maintenance req...  CAAH_BG_MATOWNER   
2  Section 16.16.030 (A) - Proper maintenance req...   CAAH_BLDG_MAINT   
3  Section 16.28.010 - Building Permit Required: ...  CAAB_BLDG_PERMIT   
4  Section 16.16.020 (B) - The provisions of this...   CAAH_BLDG_SCOPE   

  table_name expired_flag date_expired created_by date_created modified_by  \
0       NULL            N         NULL          1      03:52.6        NULL   
1       NULL            N         NULL          1      03:52.6        NULL   
2       NULL            N         NULL          1      03:52.6        NULL   
3       NULL            N         NULL          1      03:52.6        NULL   
4       NULL            N         NULL          1      03:52.6        NULL   

  date_modified fee_setup_id  
0          NULL         NULL  
1          NULL         NULL  
2          NULL         NULL  
3          NULL         NULL  
4          NULL         NULL

In [3]:
violation_types_df.tail(5)

violation_id org_id violation_type                    violation_desc  \
929          965      1          BBR36    No motions for reconsideration   
930          966      1          BBR37             Special Event Permits   
931          967      1          BBR38                  Suspension Rules   
932          969      1          BBR41  No permit issued with complaints   
933          970      1          BBR39       Sports Authority Facilities   

                                        violation_text remedial_text  \
929  The Beer Permit Board shall not entertain moti...          NULL   
930  Applicants for special event permits will be l...          NULL   
931  Upon suspension of a beer permit, the permitte...          NULL   
932  A permit shall not be issued by the Executive ...          NULL   
933  This rule is applicable only to beer sales wit...          NULL   

    table_name expired_flag date_expired created_by date_created modified_by  \
929       NULL            N         NULL      10635      04:56.0        NULL   
930       NULL            N         NULL      10635      08:59.4        NULL   
931       NULL            N         NULL      10635      10:16.9        NULL   
932       NULL            N         NULL      10635      12:53.1        NULL   
933       NULL            N         NULL      10635      16:46.8        NULL   

    date_modified fee_setup_id  
929          NULL         NULL  
930          NULL         NULL  
931          NULL         NULL  
932          NULL         NULL  
933          NULL         NULL

In [4]:
violation_types_df.columns

Index(['violation_id', 'org_id', 'violation_type', 'violation_desc',
       'violation_text', 'remedial_text', 'table_name', 'expired_flag',
       'date_expired', 'created_by', 'date_created', 'modified_by',
       'date_modified', 'fee_setup_id'],
      dtype='object')

In [5]:
# Ok. So, what I'm trying to do is reduce these 900ish to 20ish types. So let's decide which column we're going to reduce. 
# The metadata docx said the violation_id is a unique identifier. Let's test that.
print(violation_types_df.shape)

(934, 14)


In [6]:
violation_types_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 934 entries, 0 to 933
Data columns (total 14 columns):
violation_id      928 non-null object
org_id            925 non-null object
violation_type    925 non-null object
violation_desc    925 non-null object
violation_text    912 non-null object
remedial_text     788 non-null object
table_name        788 non-null object
expired_flag      788 non-null object
date_expired      788 non-null object
created_by        788 non-null object
date_created      651 non-null object
modified_by       651 non-null object
date_modified     651 non-null object
fee_setup_id      651 non-null object
dtypes: object(14)
memory usage: 102.2+ KB


In [7]:
# So, there are 934 rows, there should be 934 unique violation_id values. And, they should all be integers. 
# Let's see if they are unique.
violation_types_df.violation_id.value_counts(dropna = False)
#violation_types_df['violation_id'].value_counts(dropna = False)

-                                                                                                                                                                          119
NaN                                                                                                                                                                          6
870                                                                                                                                                                          1
938                                                                                                                                                                          1
426                                                                                                                                                                          1
883                                                                                                                          

In [8]:
# Ok. So, 119 rows have a dash in them instead of an int. Let's look at them to seee if they are dropable
violation_types_df.query('violation_id == "-"').head(5)

violation_id org_id violation_type violation_desc violation_text  \
60             -   NULL           NULL              N           NULL   
155            -   NULL           NULL              N           NULL   
238            -   NULL           NULL              N           NULL   
322            -   NULL           NULL              N           NULL   
324            -   NULL           NULL              N           NULL   

    remedial_text table_name expired_flag date_expired created_by  \
60              2    00:00.0         NULL         NULL       NULL   
155             2    00:00.0         NULL         NULL       NULL   
238             2    00:00.0         NULL         NULL       NULL   
322             2    00:00.0         NULL         NULL       NULL   
324             2    00:00.0         NULL         NULL       NULL   

    date_created modified_by date_modified fee_setup_id  
60          None        None          None         None  
155         None        None          None         None  
238         None        None          None         None  
322         None        None          None         None  
324         None        None          None         None

In [9]:
violation_types_df.query('violation_id == "-"').tail(5)

violation_id org_id violation_type violation_desc violation_text  \
697            -   NULL           NULL              N           NULL   
699            -   NULL           NULL              N           NULL   
701            -   NULL           NULL              N           NULL   
703            -   NULL           NULL              N           NULL   
705            -   NULL           NULL              N           NULL   

    remedial_text table_name expired_flag date_expired created_by  \
697             2    00:00.0         NULL         NULL       NULL   
699             2    00:00.0         NULL         NULL       NULL   
701             2    00:00.0         NULL         NULL       NULL   
703             2    00:00.0         NULL         NULL       NULL   
705             2    00:00.0         NULL         NULL       NULL   

    date_created modified_by date_modified fee_setup_id  
697         None        None          None         None  
699         None        None          None         None  
701         None        None          None         None  
703         None        None          None         None  
705         None        None          None         None

## They definitely look dropable, but how do I confirm this programatically instead of relying on a visual inspection?
####      Well, I've confirmed that the head and tail match a pattern of "dropability" (i.e.:'-, NULL, NULL, N, etch..'). 
So, I could write a loop that looks inside each row, and confirms a pattern match for each cell in that row, if it's a match then I can either mark those rows for dropping, or just drop them directly.
     But all that seems complex. Instead of checking every category of every row, maybe I could just check a subset of 'critical categories' (save some time and effort). 
##### What are my best candidates for critical categories?
   If they don't have a type, description, or text, then they are effectively uncategorizeable. That's a nicer small set of 3. (Time data is irrelevant in this context since we're dealing with violation types, not actual instances of violations.) Ok so I'll try to write a loop that checks the values in those columns for each row.

In [10]:
# I should be able to query the dropable rows and store them as a new dataframe
dropabledf = violation_types_df.query('violation_type == "NULL" and violation_desc == "N" and violation_text == "NULL"')
dropabledf.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 137 entries, 15 to 926
Data columns (total 14 columns):
violation_id      131 non-null object
org_id            137 non-null object
violation_type    137 non-null object
violation_desc    137 non-null object
violation_text    137 non-null object
remedial_text     137 non-null object
table_name        137 non-null object
expired_flag      137 non-null object
date_expired      137 non-null object
created_by        137 non-null object
date_created      0 non-null object
modified_by       0 non-null object
date_modified     0 non-null object
fee_setup_id      0 non-null object
dtypes: object(14)
memory usage: 16.1+ KB


it's interesting that I ended up with 137. That's more than the 119 that have a dash instead of a violiation_id but that's fine because the criteria for discrimination applies to the entirety of the original df. If any row is missing those three critical categories then they are uncategorizeable. And that's my primary goal with this df; to reduce it from 900ish to 20ish.

## Ok, so now I've created a df where I can put everything I want to drop. How do I drop it from the original df? 
Some sort of merge? An outermerge should eliminate all entries from the original df that are in the dropable df, since it is a proper subset of the original df. This is a defacto solution. Let's try it.

In [11]:
#help(pd.merge) makes me think that a merge may not be the right approach.

In [12]:
#let's try df.replace instead

Ok, so I'm stuck. I can drop rows by index, but I have to pass the index values as a list. Can I just get the index list from the dropable df? Let's see

In [13]:
dropabledf.index

Int64Index([ 15,  60, 155, 238, 322, 324, 380, 400, 414, 416,
            ...
            708, 896, 903, 906, 908, 914, 917, 920, 923, 926],
           dtype='int64', length=137)

In [14]:
type(dropabledf.index)

pandas.core.indexes.numeric.Int64Index

In [15]:
cleaned_df1 = violation_types_df.drop(dropabledf.index)

In [16]:
cleaned_df1.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 797 entries, 0 to 933
Data columns (total 14 columns):
violation_id      797 non-null object
org_id            788 non-null object
violation_type    788 non-null object
violation_desc    788 non-null object
violation_text    775 non-null object
remedial_text     651 non-null object
table_name        651 non-null object
expired_flag      651 non-null object
date_expired      651 non-null object
created_by        651 non-null object
date_created      651 non-null object
modified_by       651 non-null object
date_modified     651 non-null object
fee_setup_id      651 non-null object
dtypes: object(14)
memory usage: 93.4+ KB


In [17]:
cleaned_df1['violation_id'].value_counts(dropna = False)

548    1
673    1
847    1
599    1
938    1
426    1
883    1
707    1
856    1
675    1
896    1
911    1
134    1
615    1
870    1
25     1
150    1
603    1
720    1
668    1
961    1
518    1
627    1
619    1
727    1
59     1
185    1
139    1
664    1
691    1
      ..
955    1
507    1
586    1
480    1
479    1
766    1
968    1
783    1
7      1
676    1
966    1
396    1
141    1
608    1
13     1
764    1
449    1
965    1
712    1
70     1
825    1
942    1
787    1
637    1
28     1
437    1
832    1
670    1
741    1
653    1
Name: violation_id, Length: 797, dtype: int64

In [20]:
# Ok. Let's see if I can make a sub df that contains all the rows where the violation id
# isn't an integer
# cleaned_df1.query(cleaned_df1.violation_id.is_integer()) didn't work

In [22]:
# cleaned_df1['violation_id'].apply(is_integer()) didn't work

In [ ]:
# help(is_integer)

In [ ]:
# help(is_int)

In [ ]:
# help(pd.is_int)

In [ ]:
# help(pd.is_integer)

In [ ]:
# help(np.is_int)

In [ ]:
# help(isinstance)

In [23]:
#cleaned_df1.query(cleaned_df1['violation_id'].apply(type == 'int'))

In [24]:
vid_series = cleaned_df1.violation_id

In [25]:
vid_series

0                                                      1
1                                                      2
2                                                      4
3                                                      5
4                                                      6
5                                                      7
6                                                      8
7                                                      9
8                                                     10
9                                                     11
10                                                    12
11                                                    13
12                                                    14
13                                                    15
14                                                    16
16                                                    17
17                                                    18
18                             

In [ ]:
vid_series.apply(type)

In [ ]:
np.sort(vid_series)

In [ ]:
#vid_series.query('len =< 4') This didn't work

In [ ]:
#cleaned_df1['violation_id'].query(len < 4) this also didn't work

In [ ]:
#cleaned_df1.query('violation_id.len == 4') also didn't work

In [ ]:
#vids_numeric = vid_series.apply(pd.to_numeric(errors = 'coerce')) also didn't work

In [ ]:
vids_numeric = pd.to_numeric(vid_series, errors = 'coerce', downcast='integer')

In [ ]:
vids_numeric

In [ ]:
cleaned_df1.remedial_text.value_counts()

In [26]:
def conv(x):
    try:
        int(x)
        return True
    except:
        return False
    
numeric = cleaned_df1['violation_id'].apply(conv)
cleaned_df2 = cleaned_df1[numeric]

In [27]:
cleaned_df2

violation_id org_id violation_type  \
0              1      1        BANNERS   
1              2      1     BGMATOWNER   
2              4      1      BLDGMAINT   
3              5      1     BLDGPERMIT   
4              6      1      BLDGSCOPE   
5              7      1     BOARDINGRQ   
6              8      1       BOARDVAC   
7              9      1         CARROW   
8             10      1       CERTCOMP   
9             11      1        CERTOCC   
10            12      1      CHICKFARM   
11            13      1     CONSIGSALE   
12            14      1      CROWDEDFM   
13            15      1        CROWDGH   
14            16      1      DUMPDESIG   
16            17      1       DUMPSTER   
17            18      1        ECSYSHZ   
18            19      1     ECSYSHZCRT   
19            20      1      ELECEAREC   
20            21      1         ELECEQ   
21            22      1     ELECEQLIGH   
22            23      1        ELECFAC   
23            24      1     ELECSVCPAN   
24            25      1         EXPARK   
25            26      1         EXTACC   
26            27      1       EXTBAHAT   
27            28      1         EXTBAL   
28            29      1      EXTCHIMTW   
29            30      1         EXTDEC   
30            31      1        EXTDECF   
..           ...    ...            ...   
888          939      1           BBR6   
889          940      1           BBR8   
890          941      1           BBR9   
891          942      1          BBR10   
892          943      1          BBR12   
893          944      1          BBR13   
894          945      1         BBR14A   
895          946      1         BBR14B   
897          947      1          BBR15   
904          948      1          BBR16   
905          949      1          BBR17   
907          951      1          BBR19   
909          952      1          BBR21   
910          953      1          BBR22   
911          954      1          BBR24   
912          955      1          BBR26   
915          956      1          BBR28   
916          957      1          BBR29   
918          959      1          BBR30   
921          960      1          BBR31   
922          968      1          BBR40   
924          961      1          BBR32   
925          962      1          BBR34   
927          963      1          BBR33   
928          964      1          BBR35   
929          965      1          BBR36   
930          966      1          BBR37   
931          967      1          BBR38   
932          969      1          BBR41   
933          970      1          BBR39   

                               violation_desc  \
0                                     Banners   
1              Proper Maintenance Req - Owner   
2                      Proper Maintenance Req   
3                    Building Permit Required   
4                      Scope of Building Code   
5                       Boarding Requirements   
6                Boarding of Vacant Buildings   
7                 Vehicle In The Right-Of-Way   
8          Certificate of Compliance Required   
9           Certificate of Occupancy Required   
10   Chickens and Other Domestic Farm Animals   
11                           Consignment Sale   
12            Overcrowding – Unrelated Adults   
13                  Overcrowding – Group Home   
14                Dumping in permitted places   
16    Hours of Collection & Dumpster Required   
17                  Electrical System Hazards   
18            Electrical System Certification   
19                     Electrical Receptacles   
20                      Electrical Equipment    
21               Electrical Lighting Fixtures   
22             Electrical Facilities Required   
23                   Electrical Service Panel   
24    Exterior Prop Area – Vehicles – Parking   
25      Exterior Prop Area – Acces Structures   
26     Exterior of Bldgs – Basement Hatchways   
27          Exterior of Buildings - Balconies   
28    Exterior of Bldgs – C

In [29]:
cleaned_df2.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 788 entries, 0 to 933
Data columns (total 14 columns):
violation_id      788 non-null object
org_id            788 non-null object
violation_type    788 non-null object
violation_desc    788 non-null object
violation_text    775 non-null object
remedial_text     651 non-null object
table_name        651 non-null object
expired_flag      651 non-null object
date_expired      651 non-null object
created_by        651 non-null object
date_created      651 non-null object
modified_by       651 non-null object
date_modified     651 non-null object
fee_setup_id      651 non-null object
dtypes: object(14)
memory usage: 92.3+ KB
